mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+    # Number of steps to do in the scanned update method (how many anakin steps).
+    cfg.system.scan_steps = int(steps_per_rollout / anakin_act_steps)
+
+    pprint(OmegaConf.to_container(cfg, resolve=True))
+
+    # Initialise system and make learning/evaluation functions
+    (env, eval_env), q_net, q_mixer, opts, rb, learner_state, logger, key = init(cfg)
+    update = make_update_fns(cfg, env, q_net, q_mixer, opts, rb)
+
+    cfg.system.num_agents = env.num_agents
+
+    key, eval_key = jax.random.split(key)
+
+    def eval_act_fn(
+        params: FrozenDict, timestep: TimeStep, key: chex.PRNGKey, actor_state: ActorState
+    ) -> Tuple[chex.Array, ActorState]:
+        """The acting function that get's passed to the evaluator.
+        A custom function is needed for epsilon-greedy acting.
+        """
+        hidden_state = actor_state["hidden_state"]
+
+        term_or_trunc = timestep.last()
+        net_input = (timestep.observation, term_or_trunc[..., jnp.newaxis])
+        net_input = tree.map(lambda x: x[jnp.newaxis], net_input)  # add batch dim to obs
+        next_hidden_state, eps_greedy_dist = q_net.apply(params, hidden_state, net_input)
+        action = eps_greedy_dist.sample(seed=key).squeeze(0)
+        return action, {"hidden_state": next_hidden_state}
+
+    evaluator = get_eval_fn(eval_env, eval_act_fn, cfg, absolute_metric=False)
+
+    if cfg.logger.checkpointing.save_model:
+        checkpointer = Checkpointer(
+            metadata=cfg,  # Save all config as metadata in the checkpoint
+            model_name=cfg.logger.system_name,
+            **cfg.logger.checkpointing.save_args,  # Checkpoint args
+        )
+
+    # Create an initial hidden state used for resetting memory for evaluation
+    eval_batch_size = get_num_eval_envs(cfg, absolute_metric=False)
+    eval_hs = ScannedRNN.initialize_carry(
+        (jax.device_count(), eval_batch_size, cfg.system.num_agents),
+        cfg.network.hidden_state_dim,
+    )
+
+    max_episode_return = -jnp.inf
+    best_params = copy.deepcopy(unreplicate_batch_dim(learner_state.params.online))
+
+    # Main loop:
+    for eval_idx, t in enumerate(
+        range(steps_per_rollout, int(cfg.system.total_timesteps + 1), steps_per_rollout)
+    ):
+        # Learn loop:
+        start_time = time.time()
+        learner_state, (metrics, losses) = update(learner_state)
+        jax.block_until_ready(learner_state)
+
+        # Log:
+        elapsed_time = time.time() - start_time
+        eps = jnp.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+        final_metrics, ep_completed = episode_metrics.get_final_step_metrics(metrics)
+        final_metrics["steps_per_second"] = steps_per_rollout / elapsed_time
+        loss_metrics = losses
+        logger.log(
+            {"timestep": t, "epsilon": eps},
+            t,
+            eval_idx,
+            LogEvent.MISC,
+        )
```

Comment:
nit formatting to align with IQL
```suggestion
        logger.log({"timestep": t, "epsilon": eps}, t, eval_idx, LogEvent.MISC)
```